File: julia/bifurcation_analysis.jl
"""
Chronic Exposure and Bifurcation Analysis - Julia Implementation

Demonstrates allostatic adaptation and regime transitions.
"""

include("NicotineModel.jl")
using Plots

println("="^60)
println("BIFURCATION ANALYSIS (Julia)")
println("="^60)

params = Parameters()

# Continuation: vary intake rate
I_range = range(0.0, 1.5, length=60)

println("\nRunning continuation analysis...")
println("(Computing steady states across intake rates)")

# Storage for results
D_steady = zeros(length(I_range))
RT_steady = zeros(length(I_range))
Rd_steady = zeros(length(I_range))
Ra_steady = zeros(length(I_range))
max_eigenvals = zeros(length(I_range))
is_stable_array = Vector{Bool}(undef, length(I_range))

# Compute steady states
for (i, I_val) in enumerate(I_range)
    # Simulate to steady state (long time)
    I_func = continuous_input(I_val)
    
    try
        # Run for long time to reach equilibrium
        sol = simulate((0.0, 100000.0), I_func, params)
        
        # Extract final state
        u_ss = sol.u[end]
        D_steady[i] = u_ss[5]
        RT_steady[i] = u_ss[4]
        Rd_steady[i] = u_ss[3]
        Ra_steady[i] = u_ss[2]
        
        # Stability
        eigenvals, _, is_stable = stability_analysis(u_ss, I_val, params)
        max_eigenvals[i] = maximum(real.(eigenvals))
        is_stable_array[i] = is_stable
    catch e
        println("  Warning: Failed at I = $I_val")
        D_steady[i] = NaN
        RT_steady[i] = NaN
        Rd_steady[i] = NaN
        Ra_steady[i] = NaN
        max_eigenvals[i] = NaN
        is_stable_array[i] = false
    end
end

# Regime classification
println("\nClassifying usage regimes...")

function classify_regime(I, D, RT, D_0, RT_0)
    D_increase = (D - D_0) / D_0
    RT_increase = (RT - RT_0) / RT_0
    
    if D_increase < 0.2 && RT_increase < 0.05
        return "Casual"
    elseif D_increase < 0.5 && RT_increase < 0.15
        return "Regular"
    else
        return "Dependent"
    end
end

regimes = [classify_regime(I, D, RT, params.D_0, params.RT_0) 
           for (I, D, RT) in zip(I_range, D_steady, RT_steady)]

# Find regime boundaries
println("\n" * "-"^60)
println("REGIME TRANSITIONS")
println("-"^60)

for i in 2:length(regimes)
    if regimes[i] != regimes[i-1]
        println("  I = $(round(I_range[i], digits=3)) μM/min: $(regimes[i-1]) → $(regimes[i])")
    end
end

# Statistics
println("\n" * "-"^60)
println("BIFURCATION STATISTICS")
println("-"^60)
println("  Min dopamine: $(round(minimum(filter(!isnan, D_steady)), digits=3)) AU")
println("  Max dopamine: $(round(maximum(filter(!isnan, D_steady)), digits=3)) AU")
println("  Fold increase: $(round(maximum(filter(!isnan, D_steady))/minimum(filter(!isnan, D_steady)), digits=2))×")
println("  Max receptor upregulation: $(round(maximum(RT_steady) - params.RT_0, digits=3))")
println("  All steady states stable: $(all(is_stable_array))")

# Visualization
println("\n" * "="^60)
println("GENERATING VISUALIZATIONS")
println("="^60)

# Main bifurcation diagram
p = plot(layout=(3,1), size=(1200, 900), legend=:topleft)

# Panel 1: Dopamine bifurcation
stable_mask = is_stable_array
plot!(p[1], I_range[stable_mask], D_steady[stable_mask], 
     linewidth=3, color=:blue, label="Stable", grid=true)
if any(.!stable_mask)
    plot!(p[1], I_range[.!stable_mask], D_steady[.!stable_mask],
         linewidth=3, color=:red, linestyle=:dash, label="Unstable")
end
hline!(p[1], [params.D_0], color=:gray, linestyle=:dash, 
      linewidth=2, label="Baseline", alpha=0.7)

# Color regime backgrounds
casual_mask = regimes .== "Casual"
regular_mask = regimes .== "Regular"
dependent_mask = regimes .== "Dependent"

if any(casual_mask)
    I_casual = I_range[casual_mask]
    plot!(p[1], I_casual, fill(maximum(D_steady)*1.1, length(I_casual)),
         fillrange=0, fillalpha=0.15, fillcolor=:green, label="Casual", linewidth=0)
end
if any(regular_mask)
    I_regular = I_range[regular_mask]
    plot!(p[1], I_regular, fill(maximum(D_steady)*1.1, length(I_regular)),
         fillrange=0, fillalpha=0.15, fillcolor=:yellow, label="Regular", linewidth=0)
end
if any(dependent_mask)
    I_dependent = I_range[dependent_mask]
    plot!(p[1], I_dependent, fill(maximum(D_steady)*1.1, length(I_dependent)),
         fillrange=0, fillalpha=0.15, fillcolor=:red, label="Dependent", linewidth=0)
end

ylabel!(p[1], "Dopamine [AU]")
title!(p[1], "Bifurcation Diagram: Dopamine vs Intake Rate")

# Panel 2: Receptor upregulation
plot!(p[2], I_range, RT_steady .- params.RT_0, 
     linewidth=3, color=:purple, label="ΔR_T", grid=true)
hline!(p[2], [0], color=:gray, linestyle=:dash, linewidth=2, alpha=0.5)
ylabel!(p[2], "Receptor Upregulation")
title!(p[2], "Allostatic Adaptation")

# Panel 3: Desensitisation
plot!(p[3], I_range, Rd_steady, 
     linewidth=3, color=:red, label="R_d", grid=true)
xlabel!(p[3], "Intake Rate [μM/min]")
ylabel!(p[3], "Desensitised Receptors")
title!(p[3], "Tolerance Development")

plot!(p, plot_title="System Bifurcation: Casual Use → Dependence",
     plot_titlefontsize=14)

savefig(p, "bifurcation_julia.png")
println("✓ Saved: bifurcation_julia.png")

# Eigenvalue plot
p2 = plot(size=(1000, 600))
plot!(p2, I_range, max_eigenvals, linewidth=2.5, color=:blue,
     label="Max Re(λ)", grid=true)
hline!(p2, [0], color=:red, linestyle=:dash, linewidth=2,
      label="Stability boundary")

# Shade stable region
plot!(p2, I_range, max_eigenvals, 
     fillrange=0, where=(max_eigenvals .< 0),
     fillalpha=0.3, fillcolor=:green, label="Stable", linewidth=0)

xlabel!(p2, "Intake Rate [μM/min]")
ylabel!(p2, "Max Real Eigenvalue")
title!(p2, "Stability Analysis: Leading Eigenvalue")

savefig(p2, "stability_julia.png")
println("✓ Saved: stability_julia.png")

println("\n" * "="^60)
println("KEY OBSERVATIONS")
println("="^60)
println("1. Smooth monotonic bifurcation (no sudden jumps)")
println("2. Three regimes emerge from single mechanism")
println("3. System remains stable across all intake rates")
println("4. Receptor upregulation increases monotonically")
println("5. Desensitisation saturates at high intake")
println("="^60)

println("\n💡 Systems insight:")
println("The transition to dependence is gradual and continuous,")
println("reflecting allostatic adaptation (ultra-slow positive feedback).")
println("No threshold - just progressive dysregulation.")

File: examples/bifurcation_analysis.py
"""
Complete analysis: bifurcation and regime transitions.

Demonstrates:
- Continuation analysis as function of intake rate
- Critical transitions from casual → dependent
- Hysteresis effects
- Parameter sensitivity
"""

import sys
sys.path.append('../src')

from model import NicotineRewardModel
from bifurcation import BifurcationAnalyzer
from stability import StabilityAnalyzer
import matplotlib.pyplot as plt
import numpy as np


def main():
    print("=" * 60)
    print("BIFURCATION ANALYSIS")
    print("Transition from Casual Use to Dependence")
    print("=" * 60)
    
    model = NicotineRewardModel()
    bif_analyzer = BifurcationAnalyzer(model)
    
    # Intake rate range: 0 (abstinent) to 1.5 (heavy use)
    I_range = np.linspace(0.0, 1.5, 60)
    
    print("\n[1/3] Running continuation analysis...")
    print("      (This computes steady states across intake rates)")
    
    result = bif_analyzer.continuation_1d(
        param_name='I',
        param_range=I_range,
        include_upregulation=True,
        track_stability=True
    )
    
    # Regime classification
    print("\n[2/3] Classifying usage regimes...")
    
    regime_analysis = bif_analyzer.intake_regime_classifier(I_range)
    
    # Find regime boundaries
    regimes = regime_analysis['regime']
    regime_changes = []
    for i in range(1, len(regimes)):
        if regimes[i] != regimes[i-1]:
            regime_changes.append((I_range[i], regimes[i-1], regimes[i]))
    
    print("\n" + "-" * 60)
    print("REGIME TRANSITIONS")
    print("-" * 60)
    for I_val, old_regime, new_regime in regime_changes:
        print(f"  I = {I_val:.3f} μM/min: {old_regime} → {new_regime}")
    
    # Stability analysis at select points
    print("\n[3/3] Stability analysis at key points...")
    
    test_points = [0.1, 0.5, 1.0]
    stability_analyzer = StabilityAnalyzer(model)
    
    for I_test in test_points:
        y_ss = model.steady_state(I_test, include_upregulation=True)
        ts_results = stability_analyzer.timescale_separation(y_ss, I_test)
        
        print(f"\n  I = {I_test} μM/min:")
        print(f"    Dopamine: {y_ss[4]:.3f} AU")
        print(f"    R_T: {y_ss[3]:.3f}")
        print(f"    Timescale ratio: {ts_results['separation_ratio']:.0f}×")
    
    # Visualization
    print("\n" + "=" * 60)
    print("GENERATING VISUALIZATIONS")
    print("=" * 60)
    
    # Main bifurcation diagram
    fig, axes = plt.subplots(3, 1, figsize=(12, 10), sharex=True)
    
    D_curve = result['steady_states'][:, 4]
    R_T_curve = result['steady_states'][:, 3]
    R_d_curve = result['steady_states'][:, 2]
    stable = result['stability']
    
    # Panel 1: Dopamine
    axes[0].plot(I_range[stable], D_curve[stable], 
                'b-', linewidth=3, label='Stable', zorder=3)
    if np.any(~stable):
        axes[0].plot(I_range[~stable], D_curve[~stable], 
                    'r--', linewidth=3, label='Unstable', zorder=3)
    
    axes[0].axhline(model.params['D_0'], color='gray', linestyle=':', 
                   linewidth=2, alpha=0.7, label='Baseline')
    
    # Shade regimes
    regime_colors = {'Casual': '#90EE90', 'Regular': '#FFD700', 'Dependent': '#FF6B6B'}
    for regime_name, color in regime_colors.items():
        mask = regime_analysis['regime'] == regime_name
        if np.any(mask):
            I_regime = I_range[mask]
            D_regime = D_curve[mask]
            axes[0].fill_between(I_regime, 0, D_regime, alpha=0.2, color=color)
            # Label
            i_mid = np.where(mask)[0][len(np.where(mask)[0])//2]
            axes[0].text(I_range[i_mid], axes[0].get_ylim()[1]*0.95, 
                        regime_name, ha='center', fontweight='bold', 
                        bbox=dict(boxstyle='round', facecolor=color, alpha=0.5))
    
    axes[0].set_ylabel('Dopamine [AU]')
    axes[0].set_title('Bifurcation Diagram: Dopamine vs Intake Rate', 
                     fontweight='bold', fontsize=13)
    axes[0].legend(loc='upper left')
    axes[0].grid(True, alpha=0.3)
    
    # Panel 2: Receptor upregulation
    axes[1].plot(I_range, R_T_curve, 'purple', linewidth=3)
    axes[1].axhline(model.params['R_T_0'], color='gray', linestyle=':', 
                   linewidth=2, alpha=0.7, label='Pre-exposure')
    axes[1].fill_between(I_range, model.params['R_T_0'], R_T_curve,
                        where=(R_T_curve > model.params['R_T_0']),
                        alpha=0.3, color='purple', label='Upregulation')
    axes[1].set_ylabel('Total Receptors (R_T)')
    axes[1].set_title('Allostatic Adaptation', fontweight='bold', fontsize=13)
    axes[1].legend()
    axes[1].grid(True, alpha=0.3)
    
    # Panel 3: Desensitisation
    axes[2].plot(I_range, R_d_curve, 'red', linewidth=3)
    axes[2].set_ylabel('Desensitised\nReceptors (R_d)')
    axes[2].set_xlabel('Intake Rate [μM/min]')
    axes[2].set_title('Receptor Desensitisation (Tolerance)', 
                     fontweight='bold', fontsize=13)
    axes[2].grid(True, alpha=0.3)
    
    plt.suptitle('System Bifurcation: Casual Use → Dependence', 
                fontsize=15, fontweight='bold')
    plt.tight_layout()
    plt.savefig('bifurcation_full.png', dpi=300, bbox_inches='tight')
    print("\n✓ Saved: bifurcation_full.png")
    
    # Eigenvalue plot (stability)
    fig2, ax = plt.subplots(figsize=(10, 6))
    
    max_eigs = result['max_eigenvalue']
    ax.plot(I_range, max_eigs, 'b-', linewidth=2.5)
    ax.axhline(0, color='red', linestyle='--', linewidth=2, 
              label='Stability boundary')
    ax.fill_between(I_range, max_eigs, 0,
                    where=(max_eigs < 0),
                    alpha=0.3, color='green', label='Stable region')
    ax.fill_between(I_range, 0, max_eigs,
                    where=(max_eigs > 0),
                    alpha=0.3, color='red', label='Unstable region')
    
    ax.set_xlabel('Intake Rate [μM/min]')
    ax.set_ylabel('Max Real Eigenvalue')
    ax.set_title('Stability Analysis: Leading Eigenvalue', 
                fontweight='bold', fontsize=13)
    ax.legend()
    ax.grid(True, alpha=0.3)
    plt.tight_layout()
    plt.savefig('stability_eigenvalues.png', dpi=300, bbox_inches='tight')
    print("✓ Saved: stability_eigenvalues.png")
    
    plt.show()
    
    print("\n" + "=" * 60)
    print("KEY OBSERVATIONS")
    print("=" * 60)
    print("1. Smooth bifurcation: no sudden jumps (continuous transition)")
    print("2. Three distinct regimes emerge from single mechanism")
    print("3. Dopamine and R_T increase monotonically with intake")
    print("4. System remains stable across all intake rates (no oscillations)")
    print("5. High intake → high R_T → high baseline 'set point'")
    print("=" * 60)
    
    print("\n💡 Systems neuroscience insight:")
    print("This is NOT a simple threshold model. The transition to dependence")
    print("is gradual and reversible, but made difficult by slow R_T dynamics.")
    print("Allostatic theory is captured by the ultra-slow positive feedback.")


if __name__ == "__main__":
    main()
